src/transformers/generation/utils.py

```diff
@@ -3404,184 +3439,241 @@ def _beam_search(
+            # Only the top `num_beam` sequences can be considered for the final returned sequences
+            top_num_beam_mask = torch.cat(
+                (
+                    torch.ones((num_beams), dtype=torch.bool),
+                    torch.zeros((beams_to_keep - num_beams), dtype=torch.bool),
+                ),
+                dim=0,
+            ).to(next_token_hits_stopping_criteria.device)
+            did_top_num_beams_just_finished = next_token_hits_stopping_criteria & top_num_beam_mask[None, :]
+            # To prevent these just finished sequences from being used in subsequent iterations, set their log probs
+            # to a very large negative value
+            running_topk_log_probs = topk_log_probs + next_token_hits_stopping_criteria.to(torch.float32) * -1.0e9
+
+            # e. Get running `num_beam` sequences for the next generation step
+            next_topk_indices = torch.topk(running_topk_log_probs, k=num_beams)[1]
+            running_sequences, running_beam_scores, running_beam_indices = gather_beams(
+                [topk_running_sequences, running_topk_log_probs, topk_running_beam_indices], next_topk_indices
+            )
 
-            input_ids = torch.cat([input_ids[beam_idx, :], beam_next_tokens.unsqueeze(-1)], dim=-1)
+            # f. Further process topk logits
+            # - add length penalty
+            topk_log_probs = topk_log_probs / ((cur_len + 1 - decoder_prompt_len) ** length_penalty)
+            # - make sure no scores can be added anymore if beam is full and early stopping is on
+            beams_in_batch_are_full = torch.all(is_sent_finished, axis=-1, keepdims=True) & (early_stopping is True)
+            topk_log_probs += beams_in_batch_are_full.to(torch.float32) * -1.0e9
+            # - make sure still running sequences cannot be chosen as finalized beam
+            topk_log_probs += (~did_top_num_beams_just_finished) * -1.0e9
+
+            # g. Get finalized  `num_beam` sequences for the next generation step -- combine the previous finalized
+            # data with the new finalized sequences (if any, non-finalized sequences have a very large negative score
+            # in this step), and keep the best `num_beams` sequences.
+            # Variable shape tensors that grow with generation are padded before selecting the best beams, to ensure
+            # all beams have the same length
+            padded_sequences = nn.functional.pad(sequences, (0, 1), value=pad_token_id)
+            padded_beam_indices = nn.functional.pad(beam_indices, (0, 1), value=-1)
+
+            merged_sequences = torch.cat([padded_sequences, topk_running_sequences], dim=1)
+            merged_beams = torch.cat([padded_beam_indices, topk_running_beam_indices], dim=1)
+            merged_scores = torch.cat([beam_scores, topk_log_probs], dim=1)
+            merged_is_sent_finished = torch.cat([is_sent_finished, did_top_num_beams_just_finished], dim=1)
```

Comment:
the amount of torch.cat is scary 😓 that can slow down a lot as well. If we know all the shapes in advance (AFAIK the running shapes is never different no?) we should just write to certain tensors again and again

Reply:
We can, I wrote it that way in the TF version :p 

I wrote it this was because it is more in a vanilla `torch` style, but happy to do it in static-shape (XLA) style